### PR TITLE
Fix failing to initialize if game's Data folder is just 'Data'

### PR DESCRIPTION
**Pull request by ManlyMarco**
Any Unity game can have its _Data folder renamed to just `Data` and continue working just fine. The issue is BepInEx assumes that there is always a game name prefix, which causes a startup crash.

Closes #974

## How Has This Been Tested?
Tested locally on a game with a bunch of plugins already installed. Renamed data folder to `Data` and tried to regenerate interop assemblies, which crashed on master but worked fine on this PR.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: BepInEx.Core/Paths.cs [hunk 1]
```diff
@@ -104,20 +104,31 @@ public static void SetExecutablePath(string executablePath,
 
         GameRootPath = PlatformHelper.Is(Platform.MacOS)
                            ? Utility.ParentDirectory(executablePath, 4)
-                           : Path.GetDirectoryName(executablePath);
+                           : Path.GetDirectoryName(executablePath) 
+                          ?? throw new DirectoryNotFoundException("Failed to extract GameRootPath from executablePath: " + executablePath);
```
Comment (js6pak) [inline BepInEx.Core/Paths.cs:108]: You are only applying this null check to `Path.GetDirectoryName` even though `Utility.ParentDirectory` will also return `null` in the same way. It will only return `null` for paths that don't have a parent like `/` or `C:/` so we are safe here though.
This should be tackled in a tree wide nullability PR not here.

File: BepInEx.Core/Paths.cs [hunk 2]
```diff
@@ -104,20 +104,31 @@ public static void SetExecutablePath(string executablePath,
...
         ManagedPath = managedPath ?? Path.Combine(GameDataPath, "Managed");
         BepInExRootPath = bepinRootPath ?? Path.Combine(GameRootPath, "BepInEx");
         ConfigPath = Path.Combine(BepInExRootPath, "config");
         BepInExConfigPath = Path.Combine(ConfigPath, "BepInEx.cfg");
         PluginPath = Path.Combine(BepInExRootPath, "plugins");
         PatcherPluginPath = Path.Combine(BepInExRootPath, "patchers");
         BepInExAssemblyDirectory = Path.Combine(BepInExRootPath, "core");
-        BepInExAssemblyPath = Path.Combine(BepInExAssemblyDirectory,
-                                           $"{Assembly.GetExecutingAssembly().GetName().Name}.dll");
+        BepInExAssemblyPath = Path.Combine(BepInExAssemblyDirectory, $"{Assembly.GetExecutingAssembly().GetName().Name}.dll");
```
Comment (js6pak) [inline BepInEx.Core/Paths.cs:131]: Unrelated formatting change
Sidenote: why do we expose this

File: BepInEx.Core/Paths.cs [hunk 3]
```diff
@@ -104,20 +104,31 @@ public static void SetExecutablePath(string executablePath,
 
         GameRootPath = PlatformHelper.Is(Platform.MacOS)
                            ? Utility.ParentDirectory(executablePath, 4)
-                           : Path.GetDirectoryName(executablePath);
+                           : Path.GetDirectoryName(executablePath) 
+                          ?? throw new DirectoryNotFoundException("Failed to extract GameRootPath from executablePath: " + executablePath);
+
+        if (managedPath != null && gameDataRelativeToManaged)
```
Comment (js6pak) [inline BepInEx.Core/Paths.cs:110]: How does `gameDataRelativeToManaged` work? When would it ever be the case?

**Review by js6pak — CHANGES_REQUESTED** [review]

File: BepInEx.Core/Paths.cs (hunk 3 above)
Comment (ManlyMarco) [reply BepInEx.Core/Paths.cs:109]: It's true in mono net35.

**Review by js6pak — APPROVED** [review]

File: BepInEx.Core/Paths.cs [hunk 4]
```diff
@@ -106,9 +106,20 @@ public static void SetExecutablePath(string executablePath,
...
-        GameDataPath = managedPath != null && gameDataRelativeToManaged
-                           ? Path.GetDirectoryName(managedPath)
-                           : Path.Combine(GameRootPath, $"{ProcessName}_Data");
+        if (managedPath != null && gameDataRelativeToManaged)
+        {
+            GameDataPath = Path.GetDirectoryName(managedPath);
+        }
+        else
+        {
+            GameDataPath = Path.Combine(GameRootPath, $"{ProcessName}_Data");
```
Comment (ManlyMarco) [inline BepInEx.Core/Paths.cs:115]:
```suggestion
            // According to some experiments, Unity checks whether globalgamemanagers/data.unity3d exists in the data folder before picking it.
            // 'ProcessName_Data' folder is checked first, then if that fails 'Data' folder is checked. If neither is valid, the player crashes.
            // A simple Directory.Exists check is accurate enough while being less likely to break in case these conditions change.
            GameDataPath = Path.Combine(GameRootPath, $"{ProcessName}_Data");
```
